pkg/storegateway/series_refs.go

```diff
@@ -798,11 +878,135 @@ func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
+		return
+	}
+	diffWithNextChunk := chunkOffset(firstRef) - lastSeriesChunk.segFileOffset
+	if lastSeriesChunk.length > diffWithNextChunk {
+		lastRange.refs[len(lastRange.refs)-1].length = diffWithNextChunk
+	}
+}
+
+// partitionChunks creates a slice of seriesChunkRefsRange for each range of chunks within the same segment file.
+// It may also create more ranges if there are more chunks.
+// The partitioning here should be fairly static and not depend on the actual Series() request because
+// the resulting ranges may be used for caching, and we want our cache entries to be reusable between requests.
+func partitionChunks(chks []chunks.Meta) [][]chunks.Meta {
+	const (
+		chunksRangesPerSeries = 2
+		minChunksPerRange     = 2
+	)
+	chunksPerRange := util_math.Max(minChunksPerRange, len(chks)/chunksRangesPerSeries)
+
+	ranges := make([][]chunks.Meta, 0, util_math.Min(chunksRangesPerSeries, len(chks)/chunksPerRange))
+
+	currentRangeFirstChunkIdx := 0
+	for i := range chks {
+		isDifferentSegmentFile := chunkSegmentFile(chks[currentRangeFirstChunkIdx].Ref) != chunkSegmentFile(chks[i].Ref)
+		currentRangeIsFull := i-currentRangeFirstChunkIdx >= chunksPerRange
+		atLastRange := len(ranges) == chunksRangesPerSeries-1
+
+		if isDifferentSegmentFile || (currentRangeIsFull && !atLastRange) {
+			ranges = append(ranges, chks[currentRangeFirstChunkIdx:i])
+			currentRangeFirstChunkIdx = i
+		}
+	}
+
+	ranges = append(ranges, chks[currentRangeFirstChunkIdx:])
+
+	return ranges
+}
+
+// metasToRanges converts partitioned metas to ranges of chunk refs. It excludes any ranges that do not have
+// at least one chunk which overlaps with minT and maxT
+func metasToRanges(metas [][]chunks.Meta, blockID ulid.ULID, minT, maxT int64) []seriesChunkRefsRange {
+	someMetaOverlapsWithMinTMaxT := func(gr []chunks.Meta) bool {
+		for _, m := range gr {
+			if m.MinTime <= maxT && m.MaxTime >= minT {
+				return true
+			}
+		}
+		return false
+	}
+
+	rangesWithinTime := 0
+	for _, gr := range metas {
+		if someMetaOverlapsWithMinTMaxT(gr) {
+			rangesWithinTime++
+		}
+	}
+	ranges := make([]seriesChunkRefsRange, 0, rangesWithinTime)
+
+	for gIdx, g := range metas {
+		if !someMetaOverlapsWithMinTMaxT(g) {
+			continue
+		}
+
+		chunkRefs := make([]seriesChunkRef, 0, len(g))
+		for cIdx, c := range g {
+			var chunkLen uint32
+			// We can only calculate the length of this chunk, if we know the ref of the next chunk
+			// and the two chunks are in the same segment file.
+			// We do that by taking the difference between the chunk references. This works since the chunk references are offsets in a file.
+			// If the chunks are in different segment files (unlikely, but possible),
+			// then this chunk ends at the end of the segment file, and we don't know how big the segment file is.
+			if nextRef, ok := nextChunkRef(metas, gIdx, cIdx); ok && chunkSegmentFile(nextRef) == chunkSegmentFile(c.Ref) {
+				chunkLen = uint32(nextRef - c.Ref)
```

Comment:
I think would provide more clarity if we do the diff of `chunkOffset()` of the two refs, without assuming the actual format of the ref.